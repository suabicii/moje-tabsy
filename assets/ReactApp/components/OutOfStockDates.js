import React from "react";
import dayjs from "dayjs";

function OutOfStockDates({drugList}) {
    const dates = [];

    drugList.forEach(({name, dosing, dosingMoments, quantity}) => {
        const dosesAmount = Object.keys(dosingMoments).length;
        const daysToOutOfStock = Math.floor(quantity / (dosing * dosesAmount));
        dates.push({
            drugName: name,
            day: dayjs().add(daysToOutOfStock, 'd')
        });
    });

    return (
        <ul className="list-unstyled">
            {dates.map(({drugName, day}, index) => (
                <li key={index}><strong>{drugName}:</strong> {day.format('DD-MM-YYYY')}</li>
            ))}
        </ul>
    );
}

export default OutOfStockDates;