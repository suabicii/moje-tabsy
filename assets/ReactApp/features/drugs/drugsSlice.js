import {createSelector, createSlice} from "@reduxjs/toolkit";
import {fetchData} from "../../utils/fetchData";

const drugsSlice = createSlice({
    name: 'drugs',
    initialState: [],
    reducers: {
        addDrug: (state, action) => {
            const {id, name, quantity, quantityMax, unit, dosing, dosingMoments} = action.payload;
            state.push({id, name, quantity, quantityMax, unit, dosing, dosingMoments});
        },
        removeDrug: (state, action) => {
            const id = action.payload;
            return state.filter(drug => drug.id !== id);
        },
        editDrug: (state, action) => {
            const {id, ...updates} = action.payload;
            const drugIndex = state.findIndex(drug => drug.id === id);
            if (drugIndex !== -1) {
                state[drugIndex] = {...state[drugIndex], ...updates};
            }
        },
        setDrugs: (state, action) => action.payload
    },
    sort: true
});

const {reducer, actions} = drugsSlice;
export const {addDrug, removeDrug, editDrug, setDrugs} = actions;

const defaultDrugsSelector = state => state.drugs;

export const sortedDrugsSelector = createSelector(
    defaultDrugsSelector,
    drugs => [...drugs].sort((a, b) => a.name.localeCompare(b.name))
);

export const fetchDrugs = () => async dispatch => {
    await fetchData('/api/drug-list').then(data => {
        dispatch(setDrugs(data));
    }).catch(err => {
        console.log(err);
    });
};

export default reducer;