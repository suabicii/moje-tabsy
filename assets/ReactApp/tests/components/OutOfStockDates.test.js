import React from "react";
import ReactShallowRenderer from "react-test-renderer/shallow";
import OutOfStockDates from "../../components/OutOfStockDates";
import drugs from "./fixtures/drugs";
import Mockdate from "mockdate";

beforeAll(() => {
    Mockdate.set('2020-01-01');
});

afterAll(() => {
    Mockdate.reset();
});


it('should correctly render OutOfStockDates component', () => {
    const renderer = new ReactShallowRenderer();

    renderer.render(<OutOfStockDates drugList={drugs}/>);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
});