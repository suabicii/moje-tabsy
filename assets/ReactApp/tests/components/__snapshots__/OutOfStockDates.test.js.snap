// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render OutOfStockDates component 1`] = `
<ul
  className="list-unstyled"
>
  <li>
    <strong>
      Xanax
      :
    </strong>
     
    31-01-2020
  </li>
  <li>
    <strong>
      Witamina C
      :
    </strong>
     
    21-01-2020
  </li>
  <li>
    <strong>
      Metanabol
      :
    </strong>
     
    05-01-2020
  </li>
</ul>
`;
